types/base.d.ts

```diff
@@ -99,20 +102,19 @@ export interface CallbackTestContext extends TestContext {
 	 */
 	end(): void;
 }
-export interface ContextualTestContext extends TestContext {
-	context: any;
-}
-export interface ContextualCallbackTestContext extends CallbackTestContext {
-	context: any;
-}
+
+export type GenericTestContext<T> = TestContext & T;
+export type GenericCallbackTestContext<T> = CallbackTestContext & T;
 
 export interface Macro<T> {
 	(t: T, ...args: any[]): void;
 	title? (providedTitle: string, ...args: any[]): string;
 }
 export type Macros<T> = Macro<T> | Macro<T>[];
 
-export function test(name: string, run: ContextualTest): void;
-export function test(run: ContextualTest): void;
-export function test(name: string, run: Macros<ContextualTestContext>, ...args: any[]): void;
-export function test(run: Macros<ContextualTestContext>, ...args: any[]): void;
+interface TestFunctionCore<T> {
```

Comment:
Can we think of a better name here? Technically this is not the function that does the tests, it only registers a test. So maybe something like `RegisterBase<T>`, and rename `TestFunction` `Register<T>`?

And add `export` here